### ISSUE TITLE
Large quantity of schedules/backups cause high memory usage over time

### ISSUE THREAD

**Issue by twforeman** [heptio/ark#780]
We are creating schedules for each namespace - around 100 of them. Each schedule runs a backup every hour, with a ttl of 7 days. The schedules are offset by one minute each, so at most two run at the same minute.

Current backup counts:
Completed: 13286
Failed: 949

After running for about 5 days the memory usage of the ark container is 4gb and continues to increase.

Setting a memory limit of 1gb causes the container to restart with OOM errors after about 12 hours. Upping the limit to 1.5gb increases the time between restarts, but it still OOM crashes after a while. The OOM crashes cause failed backups.

Possible memory leak?

The logs show failures due to the OOM issue, but don't really indicate any issues.

Sample schedule and backup: https://gist.github.com/twforeman/f1e2289360a5c5c29701b284f7928ea4

The failed backups have no logs.

**Environment:**

- Ark version (use `ark version`):
 > ark version
Version: v0.9.0
Git commit: v0.9.0
Git tree state: clean

- Kubernetes version (use `kubectl version`):
We are running in Openshift 3.9:
 > oc version
oc v3.9.0+191fece
kubernetes v1.9.1+a0ce1bc657
openshift v3.9.30

- Cloud provider or hardware configuration:
We are running in AWS and backups are stored in S3

- OS (e.g. from `/etc/os-release`):
NAME="Red Hat Enterprise Linux Server"
VERSION="7.5 (Maipo)"

**Comment by skriss** [comment]
Thanks for sharing this information @twforeman.  We'll need to do some investigation here to see if we can figure out what's causing the runaway memory usage.  Could you also provide the count of `DeleteBackupRequests` (these are Ark-defined CRDs)?

**Comment by rosskukulinski** [comment]
@skriss we should also take advantage of the Golang metrics that are exposed through Prometheus.  This will give memory & GC details for the go runtime.

There area few Grafana dashboards as well: https://grafana.com/dashboards/3574 and https://grafana.com/dashboards/240

**Comment by skriss** [comment]
@twforeman `kubectl get deletebackuprequests -n heptio-ark | wc -l` should do the trick

**Comment by twforeman** [comment]
Looks like `oc get deletebackuprequests.ark.heptio.com -n heptio-ark` should work (in Openshift) but it returns `No resources found`

The other CRDs return values, so I think there aren't any deletebackuprequests

**Comment by skriss** [comment]
OK, well that's a good sign.  We had issues in previous releases with a huge number of those CRDs piling up, so just wanted to rule that out as an issue.

**Comment by ncdc** [comment]
@twforeman

We'll probably need to enable pprof support for the ark server and get some heap dumps from it. I can resurrect my PR for that and we can do some local simulations where we create ~15k backups and see how Ark behaves.

Does the average backup include PV snapshots? Could you possibly share an example backup yaml (redact any sensitive information, if applicable) - `ark backup get <name> -o yaml`?

Do you have any `restores`?

**Comment by twforeman** [comment]
We are not currently backing up PVs. Mainly because the back end in the cluster is GlusterFS and I can't be bothered to figure out how to make that work. :) It's our test cluster. In prod we are using EBS volumes and I'll be backing those up.

```
 > ark get backup recommendations-test-20180818040501 -o yaml
apiVersion: ark.heptio.com/v1
kind: Backup
metadata:
  creationTimestamp: 2018-08-18T04:05:02Z
  labels:
    ark-schedule: recommendations-test
  name: recommendations-test-20180818040501
  namespace: heptio-ark
  resourceVersion: "79137277"
  selfLink: /apis/ark.heptio.com/v1/namespaces/heptio-ark/backups/recommendations-test-20180818040501
  uid: e180f6e4-a29b-11e8-93d6-0e811deaacd0
spec:
  excludedNamespaces: null
  excludedResources: null
  hooks:
    resources: null
  includeClusterResources: null
  includedNamespaces:
  - recommendations-test
  includedResources: null
  labelSelector: null
  ttl: 168h0m0s
status:
  completionTimestamp: 2018-08-18T04:05:07Z
  expiration: 2018-08-25T04:05:04Z
  phase: Completed
  startTimestamp: 2018-08-18T04:05:04Z
  validationErrors: null
  version: 1
  volumeBackups: null
```

We haven't had to do any restores yet, other than some testing to make sure it actually worked. It did. :)

**Comment by ncdc** [comment]
It appears that the some of the underlying data structures used by the gzip Writer that we create for each backup are persisting in the heap instead of being garbage collected:

```
heap profile: 182: 86310864 [1738: 205693080] @ heap/1048576
115: 76308480 [115: 76308480] @ 0x1269e71 0x1275961 0x1aef655 0x1aefa5c 0x1de5998 0x1f675e7 0x1f6643d 0x1f83c8e 0x1f65cb1 0x1f65bbb 0x1f83cea 0x14237e4 0x14230cd 0x1422ffd 0x1f80d58 0x105b7e1
#	0x1269e70	compress/flate.NewWriter+0x30												/usr/local/Cellar/go/1.10.3/libexec/src/compress/flate/deflate.go:667
#	0x1275960	compress/gzip.(*Writer).Write+0x3b0											/usr/local/Cellar/go/1.10.3/libexec/src/compress/gzip/gzip.go:191
#	0x1aef654	github.com/heptio/ark/vendor/github.com/sirupsen/logrus.Entry.log+0x354							/Users/andy/go/src/github.com/heptio/ark/vendor/github.com/sirupsen/logrus/entry.go:114
#	0x1aefa5b	github.com/heptio/ark/vendor/github.com/sirupsen/logrus.(*Entry).Info+0xab						/Users/andy/go/src/github.com/heptio/ark/vendor/github.com/sirupsen/logrus/entry.go:141
#	0x1de5997	github.com/heptio/ark/pkg/backup.(*kubernetesBackupper).Backup+0x3c7							/Users/andy/go/src/github.com/heptio/ark/pkg/backup/backup.go:236
#	0x1f675e6	github.com/heptio/ark/pkg/controller.(*backupController).runBackup+0x3d6						/Users/andy/go/src/github.com/heptio/ark/pkg/controller/backup_controller.go:365
#	0x1f6643c	github.com/heptio/ark/pkg/controller.(*backupController).processBackup+0x5fc						/Users/andy/go/src/github.com/heptio/ark/pkg/controller/backup_controller.go:279
#	0x1f83c8d	github.com/heptio/ark/pkg/controller.(*backupController).(github.com/heptio/ark/pkg/controller.processBackup)-fm+0x3d	/Users/andy/go/src/github.com/heptio/ark/pkg/controller/backup_controller.go:102
#	0x1f65cb0	github.com/heptio/ark/pkg/controller.(*backupController).processNextWorkItem+0xd0					/Users/andy/go/src/github.com/heptio/ark/pkg/controller/backup_controller.go:191
#	0x1f65bba	github.com/heptio/ark/pkg/controller.(*backupController).runWorker+0x2a							/Users/andy/go/src/github.com/heptio/ark/pkg/controller/backup_controller.go:178
#	0x1f83ce9	github.com/heptio/ark/pkg/controller.(*backupController).(github.com/heptio/ark/pkg/controller.runWorker)-fm+0x29	/Users/andy/go/src/github.com/heptio/ark/pkg/controller/backup_controller.go:165
#	0x14237e3	github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1+0x53					/Users/andy/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
#	0x14230cc	github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil+0xbc						/Users/andy/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
#	0x1422ffc	github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait.Until+0x4c						/Users/andy/go/src/github.com/heptio/ark/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
#	0x1f80d57	github.com/heptio/ark/pkg/controller.(*backupController).Run.func2+0x67							/Users/andy/go/src/github.com/heptio/ark/pkg/controller/backup_controller.go:165
```

I'm continuing to investigate.

**Comment by ncdc** [comment]
This has been fixed in master. In 0.9.x, commenting out https://github.com/heptio/ark/blob/release-0.9/pkg/backup/item_backupper.go#L324 appears to make this leak go away. I'd ideally like to fix the leak in 0.9.x, but I'm not yet sure what's holding on to the references here.

**Comment by ncdc** [comment]
I've found out what is causing the leak. Our base go-plugin config is this:

https://github.com/heptio/ark/blob/release-0.9/pkg/plugin/manager.go#L44-L48

`Managed: true` is the source of the problem. When this is set, it instructs the go-plugin library to store a reference to all plugin clients ever instantiated. The expectation is that you'll eventually call the library's `CleanupClients()` function. We do call it, but unfortunately in 0.9.x it's a deferred call that only ever executes at server shutdown. In master, we don't set `Managed: true` and we handle killing plugin clients at the appropriate times.

In 0.9.x, to fix this, we'll need to set `Managed: false` and make sure that we're killing the plugin clients at all the right times. I think the only change required is to adjust `manager.CleanupClients` to kill all clients instead of calling go-plugin's function to clean up all managed clients.

**Comment by skriss** [comment]
This was fixed by #797 which has been released in v0.9.4+
